[Traffic Manager] Adding new preview api-version for Traffic Manager. This adds ONE new feature to Traffic Manager: HeatMap.

hrkulkarMsft opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [X] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [X] If applicable, the PR references the bug/issue that it fixes.
- [X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [X] My spec meets the review criteria:
  - [X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

hrkulkarMsft committed:
[Traffic Manager] Adding a preview API version for Ignite. This api-version includes a new nested resource: Microsoft.Network/trafficManagerProfiles/heatMaps. This resource will be used to represent a heatmap of a Traffic Manager profile's traffic across a world map based off of the traffic's latency and volume.

[Traffic Manager] Adding new feature to preview API: APP RUM. Also fixing some heatmap specifications.

hrkulkarMsft committed:
Initial commit

msftclas commented:
@hrkulkarMsft,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

anuchandy commented:
@ravbhatnagar this is new new API version, please review and sign-off.

anuchandy commented:
@hrkulkarMsft please update `readme.md` file to have a tag for this new api-version.

hrkulkarMsft commented:
Should the readme.md still try to build non-preview api? Or should I add?


``` yaml
openapi-type: arm
tag: package-2017-09-preview
```

### Tag: package-2017-09-preview

These settings apply only when `--tag=package-2017-09-preview` is specified on the command line.

``` yaml $(tag) == 'package-2017-09-preview'
input-file:
- Microsoft.Network/2017-09-01-preview/trafficmanager.json
```

hrkulkarMsft committed:
Fixing readme.md.
Forcing git to rename HeatMap-Get.json to HeatMap-GET.json
Fixing examples validation and json to match examples.
Adding query parameters to HeatMap. Removing endpoint identifying
properties. Must be accessed through additional GET.
Removing RealUserMetrics API. Needs separate review.
Adding query parameters for HeatMap.

anuchandy commented:
@hrkulkarMsft, the readme file seems correct, it seems CI is failing because from the swagger you are referring `HeatMap-GET.json` but in the file system the file name is `HeatMap-Get.json`, please use the same casing in both places.

anuchandy commented:
If you already have autorest installed I would recommend to run it locally against your swagger to catch linter errors, and fix them.

Assuming you are in the root folder of local clone of rest-api-spec, you can run this:

```
autorest --validation --azure-validator --message-format=json --input-file=./specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json
```

hrkulkarMsft commented:
Thanks, my validator wasn't throwing that warning because git hadn't renamed the case remotely to match my local branch.

For Operations API -- our parent RP (Microsoft.Network) implements this should we still reference it in our Swagger?

anuchandy commented:
For `Operations API` no need to reference to the parent network `swagger`. @ravbhatnagar please note that linter shows RPCViolation saying `Operations API` is missing but the same operation is implemented in the parent RP https://github.com/Azure/azure-rest-api-specs/tree/current/specification/network/resource-manager.

anuchandy commented:
To repro semantic and model validation locally

1. install oav
     `npm install -g oav`
2. run

```
oav validate-spec ./specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json
oav validate-example ./specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json 
```

anuchandy commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -684,6 +684,187 @@
+              "$ref": "#/definitions/CloudError"
+            }
+          }
+        },
```
Is this a long running operation? if yes mark it so using ` "x-ms-long-running-operation": true`

hrkulkarMsft replied:
No, it should be quite short.

anuchandy replied:
thanks for confirming

anuchandy commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -684,6 +684,187 @@
+              "$ref": "#/definitions/CloudError"
+            }
+          }
+        },
```
Is this a long running operation? if yes mark it so using ` "x-ms-long-running-operation": true`

hrkulkarMsft replied:
No, it should be short-running.

anuchandy replied:
thanks for confirming

hrkulkarMsft commented:
Thanks Anu, I was able to successfully run these validations this time.

anuchandy commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -684,6 +684,187 @@
+            "required": false,
+            "schema": {
+              "type": "string",
+              "enum": [""]
```
curious what exactly this `enum` is? if not used please remove.

hrkulkarMsft replied:
Removed this.

anuchandy commented:
@hrkulkarMsft regarding `TrackedResourceListByImmediateParent` RPCViolation - you mentioned that `in the case of this particular resource -- it could be very large`, if i understand correctly the number of HeapMap child resources associated with an instance of traffic manager profile resource can be huge. If that is the case, this could be a paged collection right? which can be expressed using  `x-ms-pageable` extension, https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/creating-swagger.md#Paging-x-ms-pageable.

hrkulkarMsft commented:
@anuchandy Sorry, I should have been more specific. The payload can return close to 4MB in our preview version -- this will be right under ARM limit to keep it from needing pagination. If we decide to have multiple heatMap resources at some point then (i.e.: a history of a Profile's heatMap), I would imagine that a List operation could get large.

anuchandy commented:
@hrkulkarMsft thanks for clarifying, got it so there will be only one `HeapMap` child instance associated with a traffic manager profile instance, i.e. this is not a collection. I would let Gaurav comment on this. @ravbhatnagar we have `RPCViolation` warning in this case `TrackedResourceListByImmediateParent`, please share your thoughts.

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Network/trafficManagerRealUserMetricsKey": {
```
What is this key? Is it a secret? Shouldn't this be obtained via a POST?

ravbhatnagar replied:
The API signature is not correct for a GET or a PUT. it needs a segment for the resource name.

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+    },
+    "EndpointProperties": {
+      "properties": {
+        "targetResourceId": {
```
How is this different from the "id" property in "Resource" model?

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+          "format": "int64",
+          "description": "The minimum number of endpoints that must be available in the child profile in order for the parent profile to be considered available. Only applicable to endpoint of type 'NestedEndpoints'."
+        },
+        "geoMapping": {
```
Is this a fixed finite set? Consider exposing this list through an API or make an enum. Or is this list same as the supported azure locations?

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/Endpoint"
```
Are all properties in the model patchable?

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/trafficmanagerprofiles": {
```
camelcase resource type name -> trafficManagerProfiles

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+          "200": {
+            "description": "The Traffic Manager Profile was deleted successfully.",
+            "schema": {
+              "$ref": "#/definitions/DeleteOperationResult"
```
there should not be  a response body for delete operation

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+            "in": "body",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/Profile"
```
All properties are patchable?

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+        "endpoints": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/Endpoint"
```
Are you sure you want to return the entire Endpoint resource in the response of a Profile? You can just return the resource Ids for the endpoints.

ravbhatnagar replied:
You will have to do a linked access check to make sure user has access to these resources (endpoints).

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+          "format": "int64",
+          "description": "The monitor interval for endpoints in this profile. This is the interval at which Traffic Manager will check the health of each endpoint in this profile."
+        },
+        "timeoutInSeconds": {
```
This should be ISO8601 format

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+          "type": "string",
+          "description": "The path relative to the endpoint domain name used to probe for endpoint health."
+        },
+        "intervalInSeconds": {
```
ISO 8601 format datetime?

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -0,0 +1,1446 @@
+          "type": "string",
+          "description": "The ARM Resource ID of this Traffic Manager endpoint."
+        },
+        "endpointId": {
```
Is this not the arm resource id of the endpoint

ravbhatnagar reviewed (changes requested):
@hrkulkarMsft - please take a look at this feedback.

ravbhatnagar commented on specification/trafficmanager/resource-manager/Microsoft.Network/2017-09-01-preview/trafficmanager.json:
```diff
@@ -1106,6 +1287,123 @@
+      "description": "Class representing a Traffic Manager HeatMap properties."
+    },
+    "HeatMapEndpoint": {
+      "properties": {
```
lets keep just resourceid

hrkulkarMsft commented:
@ravbhatnagar : Submitted a new iteration in the review with the changes we discussed.


•	{heatMapsType}: can now only be “default” in value.
•	Removed endpoint-identifying properties.
•	Added query parameters.
•	Removed App Rum to get HeatMap through for now, and will open separate review.

I’ll take note of the general API feedback to fix, but for this iteration would it be possible to only fix the HeatMap implementation, and then update the existing API comments later?

ravbhatnagar commented:
@hrkulkarMsft - sounds good.

anuchandy commented:
RubyCoden is failing https://travis-ci.org/Azure/azure-rest-api-specs/jobs/268142263, checking with the codegen owners.

hrkulkarMsft commented:
Hey, is there anything I need to do to unblock codegen?

anuchandy commented:
hi @hrkulkarMsft sorry for the delay. No action required from your side now, we are tracking the code-gen issue here https://github.com/Azure/azure-sdk-for-ruby/issues/944. Merging this PR.

@ravbhatnagar as i mentioned above though there is a RPCViolation `OperationsAPIImplementation` reported by linter, the API is implemented in the parent Network swagger.

hrkulkarMsft commented:
Cool, thanks Anu! Changed the PR to better reflect these changes since the other feature was moved to a separate PR.

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-node

AutorestCI commented:
No modification for AutorestCI/azure-sdk-for-python
